native/src/consensus/mod.rs

```diff
@@ -0,0 +1,190 @@
+    };
+
+    let pgaddress = pg::Address::new();
+    pgaddress.create(&conn);
+    for source in sources {
+        pgaddress.input(&conn, AddrStream::new(GeoStream::new(Some(source)), context.clone(), args.error_sources.clone()));
+    }
+    pgaddress.index(&conn);
+
+    let mut source_map: HashMap<String, Option<(f64, f64)>> = HashMap::new();
+    let rows = conn.query("SELECT source FROM address GROUP BY source", &[]).unwrap();
+    for row in rows.iter() {
+        let source: String = row.get(0);
+        source_map.insert(source, None);
+    }
+
+    let query = "
+        SELECT
+            ST_Distance(ST_SetSRID(ST_Point($2, $3), 4326), p.geom),
+            json_build_object(
+                'id', p.id,
+                'number', p.number,
+                'version', p.version,
+                'names', p.names,
+                'output', p.output,
+                'source', p.source,
+                'props', p.props,
+                'geom', ST_AsGeoJSON(p.geom)::TEXT
+            )
+        FROM
+            address p
+        WHERE
+            p.number = $1
+            AND p.source = $4
+            AND ST_DWithin(ST_SetSRID(ST_Point($2, $3), 4326), p.geom, 0.01);
+    ";
+
+    let sources: Vec<String> = source_map.keys().cloned().collect();
+    let threshold = args.threshold.unwrap_or(25);
+    let mut agreement = agreement::Agreement::new(sources.clone(), threshold);
+
+    for addr in AddrStream::new(GeoStream::new(Some(query_points)), context.clone(), args.error_query_points) {
+        for source in &sources {
+            // pull the addresses matching this address number within 1 km
+            let rows = conn.query(query, &[ &addr.number, &addr.geom[0], &addr.geom[1], &source ]).unwrap();
+
+            // populate potential_matches with db response
+            let mut potential_matches: Vec<Address> = Vec::with_capacity(rows.len());
+            for row in rows.iter() {
+                let paddr: serde_json::Value = row.get(1);
+                let paddr = Address::from_value(paddr).unwrap();
+                potential_matches.push(paddr);
+            }
+
+            // find a match using linker
+            match compare(&addr, &mut potential_matches) {
+                Some(link_id) => {
+                    let mut pmatches: Vec<Address> = potential_matches.into_iter().filter(|potential| {
+                        link_id == potential.id.unwrap()
+                    }).collect();
+
+                    match pmatches.len() {
+                        0 => continue,
+                        1 => {
+                            let paddr = pmatches.pop();
+                            let coords = match paddr {
+                                Some(point) => {
+                                    Some((point.geom[0], point.geom[1]))
+                                },
+                                None => None
+                            };
```

Comment:
[nit] IMHO `Option.map` is a clearer way to express this:
```suggestion
                            let coords = paddr.map(|p| (p.geom[0], p.geom[1]);
```

Reply:
Wow nice, didn't know this existed.